### DISCUSSION TITLE
Ternoa Node Upgrade Action Plan

### DISCUSSION THREAD

**Discussion by markopoloparadox** [capsule-corp-ternoa/ternoa-hub#21] [General]
Starting from tomorrow (2022-09-01) we will begin a compulsory node upgrade process for all the existing nodes on the Alphanet and Mainnet. This process is divided into three phases:

Phase 1:  On 2022-09-01 The latest changes from substrate (monthly-2022-09) will be merged into the ternoa node project. Once this is done I will alert you here that you can start rebuilding your binaries.

Phase 2:  From 2022-09-01 until 2022-09-11  we are going to update all our Alphanet VMs to Ubuntu 22.04 and we are going to switch the existing binaries to new ones. I would suggest you to create new  validator nodes (using the old binary) on Alphanet and try to do the upgrade yourself. This should prepare you for the next phase.

Phase 3: From 2022-09-12 until 2022-09-25 we are going to update all our Mainnet VMs to Ubuntu 22.04 and we are going to switch existing binaries to new ones. This part is compulsory for all validators to ensure that the integrity of the network is not compromised.

If no problems arise, we should be able to stick with the plan but in case the substrate commit from 2022-09-01 is faulty we will delay all the phases until the issues are solved. 

QA:
**How does the upgrade process look like?**
Mostly it's just a matter or stopping the node, switching the binary and restarting the node. In our case, we will try to do the full process and instead of swapping the binary we will run competently new VMs with the new binaries. More about this [here](https://wiki.polkadot.network/docs/maintain-guides-how-to-upgrade)

**If I have a question where should I ask it?**
All the questions regarding this process should be asked in this discussion. Questions on Discord and Telegram or private messages will be ignored.

**Where can I find the latest information about this?**
This discussion thread will contain all the information that you need. This means that it's your duty to check this discussion everyday in cases unexpected changes are made.

**How do you build the binary?**
Check out our readme [here](https://github.com/capsule-corp-ternoa/ternoa-node) and our [wiki](https://github.com/capsule-corp-ternoa/ternoa-node/wiki/Running-a-Validator-Node-and-Bonding-CAPS) page on how to be a validator.

cc: @peshwar9 @soufiane-bouchaara @M2613

**Comment by markopoloparadox** [comment]
The latest changes have been merged. You can now visit [Ternoa Node Repo](https://github.com/capsule-corp-ternoa/ternoa-node) and build the binary for your machine. Make sure that you change your branch from `main` to tag `v1.2.0-rc1`.

If you have docker or podman installed, we have provided dockerimages that build the binary for the following flavors:  Ubuntu 20.04/22.04, Debian 11, Arch, Fedora 35/36.

**Comment by markopoloparadox** [comment]
Hello guys :)
Testing on Alphanet showed no issues so this means that we are doing the node upgrade on Mainnet. One important thing to mention is that now you need to pass the following flag `--state-pruning archive`  to your node otherwise it won't store all the block states. Previous this was implied when the `--validator` flag was passed but not it needs to be manually set. This flag is optional so you don't need to pass it if you don't want.

**Reply by t0sma86** [reply]
Can you write an example please with the new flag ?
Regarding the branch for building binary, we keep v1.2.0-rc1 or main ? Thanks

**Reply by LoulouCrypto** [reply]
When did we have to make the upgrade for mainnet ? 

the tag is V1.2.0-rc1 ? or do we wait for the final version ?

**Reply by markopoloparadox** [reply]
1. The tag that you need to use is v1.2.0-rc1.
2. The plan is to do the upgrade on Mainnet until the end of next week (2022-09-25)
3. For the flags, you can use this: 
```bash
/path/to/ternoa --name YourNodeName --chain mainnet -d /path/to/storage --validator
```
Optionally if you want to store the block states too you need to pass the following flag as well `--state-pruning archive`

**Reply by LoulouCrypto** [reply]
Hi team i have a problem compiling the bin : 
``error: failed to load manifest for workspace member `/home/validators/ternoa-node-1.2.0-rc1/node/cli`

Caused by:
  failed to load manifest for dependency `ternoa-service`

Caused by:
  failed to load manifest for dependency `alphanet-runtime`

Caused by:
  failed to load manifest for dependency `ternoa-bridge`

Caused by:
  failed to read `/home/validators/ternoa-node-1.2.0-rc1/ternoa-pallets/bridge/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
``

**Reply by LoulouCrypto** [reply]
i try with git clone and downloading the sources same thing

**Reply by t0sma86** [reply]
Hello,

Try these command lines to build your binary : 

apt update -y	
apt install build-essential git clang curl libssl-dev llvm libudev-dev make protobuf-compiler -y
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
exec bash
git clone https://github.com/capsule-corp-ternoa/chain.git && cd chain
git checkout v1.2.0-rc1
git submodule update --init --recursive
rustup update && rustup show
cargo build --locked --release
./target/release/ternoa -V

It's work for me :)

**Reply by LoulouCrypto** [reply]
I don't know how But i try a last time and it work ;) 
 3 MNSF nodes update ;) 
 
 Thx for the help

**Reply by markopoloparadox** [reply]
If you have a question unrelated to my comment write a new comment instead of replying it here.
But to answer your question we added `ternoa-pallets` as submodule inside `ternoa-node` repo (we also changed the name of the repo from `chain` to `ternoa-node`) . This means that you need to run `git submodule update --init --recursive` to get all the submodules.

All this is already documented on the home page of `ternoa-node`

**Comment by cryptokyser** [comment]
Done for CK-Validator

**Comment by LoulouCrypto** [comment]
Done for MNSF

**Comment by Orius** [comment]
![Visuel_fin_upgrade_mainnet](https://user-images.githubusercontent.com/64799905/190403243-f8684b64-34c2-4c9b-9dca-899ab81fdd41.jpg)

**Reply by markopoloparadox** [reply]
Love the image :D

**Comment by markopoloparadox** [comment]
@cryptokyser @LoulouCrypto @Orius 
Thank you guys for doing this upgrade. I know that this was quite scary but I think that we managed to do it without any issues.

**Reply by markopoloparadox** [reply]
Would be cool if you could get some kind of a badge (NFT) as you were the first guys to do a node upgrade. Will talk with the team and if I got a positive feedback I will let you know :)

**Reply by LoulouCrypto** [reply]
🙏 you are welcome 🙏

After building the bins the update was really easy 😁

**Reply by Orius** [reply]
Great idea 🙏

**Reply by Orius** [reply]
@t0sma86 and @Gitcrypte were also among the first to upgrade using the method recommended by the Polkadot Wiki. They also would merit this NFT, if the Ternoa team agrees to mint this souvenir 🙏

**Reply by Crypte** [reply]
Thank you Oliver for the mention, I confirm that on my side everything has been up to date for 24 hours (I spread the procedure over 4-5 days).  And I will also consider enabling notifications for GitHub😂

**Comment by t0sma86** [comment]
Hi, 
what happen if 50% of nodes do not make the upgrade before 2022-09-25 ? I don't see a lot of node operator make the update to 1.1.0 

Thanks

![image](https://user-images.githubusercontent.com/108297916/191092986-4f72ec25-3ad1-455e-9e0e-6eb7181322cd.png)

**Reply by markopoloparadox** [reply]
According the the tests, nothing will happen.

**Reply by LoulouCrypto** [reply]
I think thé most important IS that the 15 node elected are update .... And i think they are for the most updated

**Reply by cryptokyser** [reply]
The most important thing is that everyone (waiting or active) does at least this upgrade. It's not complicated. And a validator should know how to do this kind of thing easily. My CK-VALIDATOR node is still waiting but knowing how to do this update (pending or active) is a prerequisite. It must be done. Being a Validator isn't just about earning rewards, it comes with responsibilities and upgrading is part of that.
